babel/src/test/resources/sql/redshift.iq

```diff
@@ -523,78 +523,78 @@ select deptno, ratio_to_report(sal) over (partition by deptno) from emp;
 !}
 
 # STDDEV_POP
-select empno, stddev_pop(comm) over (order by empno rows unbounded preceding) from emp where deptno = 30 order by 1;
+select empno, stddev_pop(CAST(comm AS DECIMAL(12, 4))) over (order by empno rows unbounded preceding) from emp where deptno = 30 order by 1;
```

Comment:
Initially I had filed issue https://issues.apache.org/jira/browse/CALCITE-6324 about this problem. See the comments there as well. Indeed, in Calcite, as in Oracle, the result type of an aggregate is the same as the type of the value aggregated. 

Moreover, the Calcite implementation of aggregates also uses *internally* the same type for all calculations. Since STDDEV needs to square values, it requires double the precision not to overflow. So without changing the type here, this program would fail because of overflows for decimal values (prior to this PR such overflows were simply ignored, which is wrong).

The problem of internal overflows could be solved in two ways:
(1) change the internal implementation of all aggregation functions that could overflow to use a different precision for intermediate results. 
(2) require the user to use a type wide enough to accommodate possible overflows

Now, unfortunately the compiler cannot choose safely a type to guarantee that the results won't overflow, even for a simple aggregate such as AVG. The reason is that, although the compiler has a bound on the precision of the values aggregated, in general it has no bound on the number of rows in the table at runtime (especially for cases like incremental view maintenance, where the plans generated need to work even for future contents of the tables). So no matter what precision the compiler may choose for internal computations, this precision may be exceeded at runtime for some input data. 

This leaves the other alternative: the user should choose a precision high enough to carry all computations. Although the user may also not know the size of the tables, at least the user has some domain-specific knowledge. So this PR adopts the approach in (2), which is also compatible with the existing Calcite implementation of aggregates.

If we want to implement (1), it will require much more work, we have to modify the implementation of all Calcite aggregates, choosing an arbitrary type for intermediate results.

Reply:
Calcite does not specify anywhere the type of the results produced by an aggregate function. However, the type inference code for aggregates implies that the result type is the same as the data type. If this is correct, these results were wrong, because the scale of the results is not the same as the scale of the input data. 

As you see, many of these problems stem from the fact that there is no written spec for what Calcite should be doing.

I don't have access to an Oracle database to check. My knowledge of Oracle for aggregations comes from their published documentation, e.g.: https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/STDDEV.html

Postgres returns double if inputs are integers, or DECIMAL otherwise, a completely different behavior from Oracle or Calcite.

Reply:
I agree that this is unpleasant. However, please note that this implementation only affects the Enumerable-based implementation, which is used to evaluate the Quidem tests. Other runtimes can choose other strategies for the precision of intermediate results.

There was a discussion about this in JIRA: https://issues.apache.org/jira/browse/CALCITE-6324, where @julianhyde pointed out that the problems is indeed the strategy chosen for implementing functions such as STDDEV.

So I think this is really a separate issue. If we think that the implementation of STDDEV (and a few other aggregates in Calcite's runtime is wrong, we should file an issue about that and address it. Then we can revert these tests to use a narrower type for the result.

Reply:
The Enumerable implementation is used in production for various systems so it is not a test-only thing that's why I raised some concerns.

I don't have something more to add to this discussion. We may not fully agree on everything but that is fine; I trust you to advance this further as you see fit.

Reply:
@zabetak Now, In Calcite **REGR_SXX and similar aggregate functions return the wrong data type**  The detail please read https://issues.apache.org/jira/browse/CALCITE-4924  and I try to use another way to make the return type more reasonable by calculate.  @mihaibudiu This issue looks same as **CAST a literal to DECIMAL type return wrong result** https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-4871. If we can't resolv CALCITE-4824, Then we can't handlt the 4871

Reply:
I think that the current bug fix would solve both CALCITE-4924 and CALCITE-4871.
The result type is correctly inferred, but the cast to decimal does nothing.

I will take a look to see whether it would be easy to change the functions that compute STDDEV to use a higher precision internally. But that should probably be a separate PR.